### PR TITLE
Improve the token server so that it refreshes tokens and enforces bac…

### PR DESCRIPTION
**Pull request by mhammond**
…koff.

This got a little messier than I was expecting. The TokenserverClient now
exposes all its functionality via a TokenserverContext object which can
supply auth headers and exposes the api_endpoint. It will always return
a StorageResetError() if a new token changes the api_endpoint, forcing
consumers of this to take remedial action and construct a new
TokenserverClient to resolve the error.

It uses a RefCell<> to manage the state (so that we don't need to declare
the function for retrieving the context as mutable) and hands tracks it
in an Rc<> because implementing a Clone on the object is (a) tricky due to
the hawk credentials and (b) probably not desirable. Fetching a context will
panic if previous holders of a context haven't dropped their reference,
enforcing consumers to re-fetch the context each time they need it, because
each call may require a new token to be fetched.

Needs tests, but I'm not sure how to add them (and it didn't come with tests
in the first place which I could build from)

It introduces a new StorageResetError, which could also be used in the future
if we detect a change to the syncID, and changes BackoffError to take a
Duration instead of an f64.

+cc @kitcambridge, @eoger

### CODE REVIEW

**Comment by mhammond** [comment]
My first Rust PR :) Feel free to ask me to change anything that would make the changes more idiomatic

File: sync15-adapter/src/token.rs [hunk 1]
```diff
@@ -130,4 +148,84 @@ impl TokenserverClient {
...
+    base_url: String,
+    access_token: String,
+    key_id: String,
+    // The token (and derived info) that we fetch and hand back to consumers.
+    // While this seems odd, we use an Rc to try and ensure consumers can't
+    // hold on to it - we'll panic if we try and update it.
+    // (Also because we can't clone an TokenserverContext as hawk doesn't derive clone,
+    // but even if we could, we'd run the risk of a consumer using a stale
+    // clone. No idea if this actually makes sense though.)
+    current_context: RefCell<Option<Rc<TokenserverContext>>>,
```
Comment (kitcambridge) [inline sync15-adapter/src/token.rs]: Is there ever a case where `current_context.is_some() && current_backoff.is_some()`? If not, I'm wondering if we could formalize these states in an enum, something like:

```rust
enum TokenserverContextState {
  NotFetched,
  Current(TokenserverContext),
  BackoffUntil(SystemTime)
}
```

And keep a `RefCell` to that.

File: sync15-adapter/src/token.rs [hunk 2]
```diff
@@ -130,4 +148,84 @@ impl TokenserverClient {
...
+        TokenserverClient {
+            base_url,
+            access_token,
+            key_id,
+            current_context: RefCell::new(None),
+            current_backoff: Cell::new(UNIX_EPOCH),
+        }
+    }
+
+    pub fn get_context(&self, request_client: &Client) -> Result<Rc<TokenserverContext>> {
```
Comment (kitcambridge) [inline sync15-adapter/src/token.rs]: I started wondering if `get_context` could return `&TokenserverContext` instead of `Rc<TokenserverContext>`, but that seems really involved, there's no way to (AFAICT?) decay a `Ref<T>` into a `&T`, and it wouldn't be safe, anyway, since `RefCell` reimplements borrow checking dynamically.

Comment (kitcambridge) [inline sync15-adapter/src/token.rs]: It might make sense to have `TokenserverClient` hold on to `Client` instead, and have `TokenserverContext` hold a borrowed reference...we'll (I think?) create a long-lived `TokenserverClient` object for the lifetime of the app, so `TokenserverContext` should never outlive `TokenserverClient`.

File: sync15-adapter/src/token.rs [hunk 3]
```diff
@@ -20,91 +23,106 @@ header! { (XTimestamp, "X-Timestamp") => [ServerTimestamp] }
...
+    duration: u64,
+    hashed_fxa_uid: String,
 }
 
-/// This is really more of a TokenAuthenticator.
-pub struct TokenserverClient {
+// The TokenserverContext holds the TokenserverToken and anything derived
+// from that token. It supplies anything derived from the token server, such
+// as auth headers, the storage endpoint, the hashed fxa uid, etc.
+pub struct TokenserverContext {
```
Comment (kitcambridge) [inline sync15-adapter/src/token.rs]: What do you think of `TokenInfo` or `TokenAuthInfo` as a name? `Context` is quite the generic name...

File: sync15-adapter/src/token.rs [hunk 4]
```diff
@@ -20,91 +23,106 @@ header! { (XTimestamp, "X-Timestamp") => [ServerTimestamp] }
...
+                                     .header(XKeyID(tsc.key_id.clone()))
                                      .send()?;
 
         if !resp.status().is_success() {
             warn!("Non-success status when fetching token: {}", resp.status());
             // TODO: the body should be JSON and contain a status parameter we might need?
             debug!("  Response body {}", resp.text().unwrap_or("???".into()));
-            if let Some(seconds) = resp.headers().get::<RetryAfter>().map(|h| **h) {
-                bail!(error::ErrorKind::BackoffError(seconds));
+            // XXX - shouldn't we "chain" these errors - ie, a BackoffError could
```
Comment (thomcc) [inline sync15-adapter/src/token.rs]: Yeah, we probably should chain them. I doubt it matters much in practice, but it could help when debugging down the road.

File: sync15-adapter/src/token.rs [hunk 5]
```diff
@@ -20,91 +23,106 @@ header! { (XTimestamp, "X-Timestamp") => [ServerTimestamp] }
...
+            // have a TokenserverHttpError as its cause?
+            if let Some(ms) = resp.headers().get::<RetryAfter>().map(|h| (**h * 1000f64) as u64) {
```
Comment (thomcc) [inline sync15-adapter/src/token.rs]: I think we should consider sanity checking that `**h >= 0.0`. I think we'll panic otherwise, I could be convinced this is okay but my gut says we'd rather handle it explicitly.

Admittedly this should never happen.

File: sync15-adapter/src/token.rs (hunk 1 above)
Comment (thomcc) [reply sync15-adapter/src/token.rs]: Not a fan of this. In general using `Rc` and `RefCell` are a sign you aren't really writing idiomatic rust. In this case, I don't think this does what you want either (specifically, it won't actually allow you to prevent stale use).

We can't clone a Credentials but we actually can recreate them from the token, which we already have around. Storing the credentials is basically just an optimization that I wouldn't have done if I realized the Credential object had bad ergonomics like that. 

As I mentioned, this won't prevent using a stale context. The type you probably want is `Rc<RefCell<Option<TokenServerContext>>>`, which *would* let us update the callers context, however it would panic if we tried to do it while they were borrowing it. However this seems very likely to cause problems, for example, we have no guarantee they aren't borrowing the context when we decide to update it for them, which could cause a panic.

I'm pretty sure the right thing here probably something like what kit is proposing, and then hold onto a `RefCell<TokenServerContextState>` so that we can have interior mutability. Then, I don't think get_context should be a public method, or if it is it should return `Result<RefMut<TokenServerContext>>` (but I think it shouldn't be one).

**Comment by thomcc** [comment]
Here's a version that avoids both needing to clone the TokenserverContext and needing to use Rc. It also is impossible for a client to use a stale context (they aren't allowed to hold contexts), and so also avoids the risk of panic present on the `Rc<RefCell<Option<T>>>` version

https://gist.github.com/thomcc/bd9cf3f390346edb1b746b7edd2740a9. It applies on top of dce6e02.

I'll probably have nits after this though, and I agree with kit in so far as I think its possible that we might want to model the states more explicitly.

File: sync15-adapter/src/token.rs (hunk 1 above)
Comment (mhammond) [reply sync15-adapter/src/token.rs]: Thanks for the great feedback - most of the comments made are now stale (so I haven't replied to them all) as the new version of this uses an enum as Kit proposes, while still sticking with Thom's (very helpful) refactor to use a callback. This has actually made it quite nice IMO.

I'm still struggling a little here with a state that represents there was an error fetching the token (and specifically, how to get this error back to the caller), but I figured I'd get a new version up now for general feedback and to gather some ideas for that error state - they are fairly easy to spot with "todo" etc comments.

Another thing I gave up on was testing using traits, so suggestions for how to simply test are also invited (although it also seems reasonable we just open an issue for tests and accept a PR without them given none exist today)

File: sync15-adapter/src/token.rs (hunk 3 above)
Comment (mhammond) [reply sync15-adapter/src/token.rs]: Yeah, those names are better. My new names are quite different than before, and there's still a "context", but I think it makes more sense in the new patch (and its private anyway)

**Comment by mhammond** [comment]
I force-pushed, so github doesn't appear to make it obvious there's a change, but if @thomcc and @kitcambridge could look over it again I'd appreciate it!

File: sync15-adapter/src/token.rs [hunk 6]
```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
...
+    #[test]
+    fn test_something() {
+        let client = Client::builder().timeout(Duration::from_secs(30)).build().expect("can't build client");
+        let tsc = TokenserverClient::new(String::from("base_url"),
+                                         String::from("access_token"),
+                                         String::from("key_id"));
+
+        // TODO: make this actually useful!
+        let _e = tsc.api_endpoint(&client).expect_err("should fail");
+        // XXX - (a) it's an unexpected error rather than the original, and
```
Comment (thomcc) [inline sync15-adapter/src/token.rs]: I think if we care what it says (even in tests) we shouldn't use `UnexpectedError` for it.

File: sync15-adapter/src/token.rs [hunk 7]
```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
...
+                // it should be impossible to get here.
+                panic!("Can't be in NoToken state after advancing");
+            }
+            TokenState::Token(ref token_context) => {
+                // make the call.
+                func(token_context)
+            }
+            TokenState::Failed(_, _) => {
+                // XXX - todo - work out how to return the value???
+                bail!(error::unexpected("todo - work out how to return the actual error?"))
```
Comment (thomcc) [inline sync15-adapter/src/token.rs]: `TokenState::Failed(e, _) => return Err(e)` or something like that doesn't work?

File: sync15-adapter/src/token.rs [hunk 8]
```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
...
+            access_token,
+            key_id,
+            current_state: RefCell::new(TokenState::NoToken),
+        }
+    }
+
+    // Attempt to fetch a new token and return a new state reflecting that
+    // operation. If it worked a TokenState::Token state will be returned, but
+    // errors may cause other states.
+    fn fetch_token(&self, request_client: &Client, previous_endpoint: Option<String>) -> TokenState {
```
Comment (thomcc) [inline sync15-adapter/src/token.rs]: Nit: We only need ownership of `previous_endpoint` in error cases, so it would be better if this were `Option<&str>` (or `&Option<String>`, but the first is more idiomatic and usually easier to work with).

File: sync15-adapter/src/token.rs [hunk 9]
```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
...
+                        }
+                    },
+                    None => {
+                        // Never had an api_endpoint in the past, so this is OK.
+                        TokenState::Token(tc)
+                    }
+                }
+            },
+            Err(e) => {
+                match error::Error::from("error!") {
```
Comment (thomcc) [inline sync15-adapter/src/token.rs]: I think this should be matching on `e`. As it is we'll have `error::Error(error::ErrorKind::Msg("error!"))` or something like that, so there's no reason to match.
